lib/srv/server/cloudwatcher.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ec2/ec2iface"
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/cloud"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+// EC2Instances is
+type EC2Instances struct {
+	Region    string
+	Document  string
+	Instances []*ec2.Instance
+}
+
+type Watcher struct {
+	// InstancesC can be used to consume
+	InstancesC chan EC2Instances
+
+	fetchers      []*ec2InstanceFetcher
+	fetchInterval time.Duration
+	ctx           context.Context
+	cancel        context.CancelFunc
+}
+
+func (w *Watcher) Run() {
+	ticker := time.NewTicker(w.fetchInterval)
+	defer ticker.Stop()
+	for {
+		for _, fetcher := range w.fetchers {
+			inst, err := fetcher.GetEC2Instances(w.ctx)
+			if err != nil {
+				log.Error("Failed to fetch EC2 instances: ", err)
+				continue
+			}
+			select {
+			case w.InstancesC <- *inst:
+			case <-w.ctx.Done():
+			}
+		}
+		select {
+		case <-ticker.C:
+			continue
+		case <-w.ctx.Done():
+			return
+		}
+	}
+}
+
+func (w *Watcher) Stop() {
+	w.cancel()
+}
+
+func NewCloudServerWatcher(ctx context.Context, matchers []services.AWSMatcher, clients cloud.Clients) (*Watcher, error) {
+	cancelCtx, cancelFn := context.WithCancel(ctx)
+	watcher := Watcher{
+		fetchers:      []*ec2InstanceFetcher{},
+		ctx:           cancelCtx,
+		cancel:        cancelFn,
+		fetchInterval: time.Minute,
+		InstancesC:    make(chan EC2Instances),
```

Comment:
maybe create a buffered channel otherwise both producers and consumers have to sync for each discovered ec2 instance